make binarize process PAGE-XML (with AlternativeImage):

bertsky opened the pull request:
- detect whether input file was image/* or PAGE-XML
- for image file, also generate a skeleton PAGE-XML
  (with AlternativeImage for the binarized result)
- for PAGE-XML file, detect namespace (and prefix),
  then check whether AlternativeImage already exists
  or find imageFilename (both on disk and as METS ID),
  process image and add AlternativeImage for result
- reference all new files (image under OCR-D-IMG-BIN,
  PAGE-XML under output fileGrp) in METS
- encapsulate, add pointers for improvement of bashlib
  in core
- improve documentation

Commit:
make binarize process PAGE-XML (with AlternativeImage):

- detect whether input file was image/* or PAGE-XML
- for image file, also generate a skeleton PAGE-XML
  (with AlternativeImage for the binarized result)
- for PAGE-XML file, detect namespace (and prefix),
  then check whether AlternativeImage already exists
  or find imageFilename (both on disk and as METS ID),
  process image and add AlternativeImage for result
- reference all new files (image under OCR-D-IMG-BIN,
  PAGE-XML under output fileGrp) in METS
- encapsulate, add pointers for improvement of bashlib
  in core
- improve documentation

bertsky commented:
Fixes #3.

wrznr commented on ocrd-olena-binarize:
```diff
@@ -1,28 +1,254 @@
+
+function call_scribo {
+    # Binarize as PBM
+    scribo-cli "$1" "$2" "${3}.pbm"
```
The problem here is that `scribo-cli`, which is a shell wrapper itself, does not wrap anything but the Sauvola-based algorithms. Wolf and Kim and friends are not available if you fall back to `scribo-cli`. Two options:
1. Modify `scribo-cli` (via `scribo-cli.in` in the Olena repo, not recommended since a merge of a corresponding PR is unlikely to happen)
2. Fall back to the individual processors which are built in `scribo/src/binarization` if `HAVE_MAGICKXX` is set (via `configure`).

bertsky replied:
I don't think so. The use of `scribo-cli` is what we already had before this PR, and it worked for `wolf` and `kim` as well.

Here is why (from `olena-2.1/scribo/src/scribo-cli`):
```bash
*)
            cmd_bin=`echo $cmd | sed -e 's/-/_/g'`
            if test -f $bdir/$cmd_bin; then
                $bdir/$cmd_bin $@
            else
                echo "Error: Invalid command!"
                usage
            fi
```
So this simply delegates to `local/libexec/scribo/{wolf,kim,niblack,otsu}`.

wrznr replied:
You are absolutely right. It is only the CLI documentation which is restricted to `sauvola`. Sorry!

Commit:
fix order of namespace/prefix detection and preserve vs update

bertsky commented:
Just made a minor modification: if one wishes to run with `PRESERVE_NAMESPACE=0` (i.e. update the namespace of the output to 2019), then the namespace and prefix were set too late.

Maybe I should expose `PRESERVE_NAMESPACE` as  a parameter, say `update-namespace` (with inverse logic)?

Commit:
added xmlstarlet to deps-ubuntu

Commit:
fallback to input ID if image not found in METS, and use parsed mets_file param

Commit:
use pre-release ocrd; resync travis/Makefile

Commit:
make output fileGrp for images explicit, update tests

bertsky commented:
Wait, I forgot to respect `Border` if input is PAGE-XML but no `AlternativeImage` exists yet!

Commit:
fix comments, add Border cropping:

- always reset IFS after customization to avoid
  interference in unexpected places
- when processing a PAGE-XML file which does not
  already have an AlternativeImage, but contains
  Border/PrintSpace, first crop that image before
  binarization
- add "cropped" to comments in that case, and add
  existing commments in the other case

Commit:
fix adding pageId for output PAGE (inverse condition)

Commit:
fix errexit, ns_prefix can be empty (which is false in xmlstarlet)

Commit:
fallback for out_id in case in_id does not contain input_file_grp

Commit:
improve tests:

- fail immediately if single test not ok
- also check for PAGE result
- add another suite with PAGE input:
  check for PAGE result and cropped image

bertsky commented:
So this is it for now. I will make a separate PR using the new bashlib API. 

@kba @wrznr, what do you think? Should the namespace be updated to 2019 in the output, or preserved, or control that via a parameter?
